File: Unidad 3/Ejercicios/MetodoDeComposicion.R
generar_composicion <- function(n, p1, lambda1, lambda2) {
  
  X <- numeric(n)
  
  set.seed(123) 

  U <- runif(n)

  for (i in 1:n) {
    if (U[i] < p1) {
      X[i] <- rexp(1, rate = lambda1)
    } else {
      X[i] <- rexp(1, rate = lambda2)
    }
  }
  
  return(X)
}

N <- 10000          
P1 <- 0.5           
P2 <- 1 - P1        
LAMBDA1 <- 0.5      
LAMBDA2 <- 2        

muestras_comp <- generar_composicion(N, P1, LAMBDA1, LAMBDA2)

print("Primeras 10 muestras generadas:")
print(head(muestras_comp, 10))

hist(
  muestras_comp,
  breaks = 50,
  freq = FALSE,
  main = "Histograma de Muestras (Método de Composición)",
  xlab = "Valores generados",
  col = "lightblue",
  border = "black"
)

x_vals <- seq(min(muestras_comp), max(muestras_comp), length.out = 500)
densidad_teorica <- P1 * dexp(x_vals, rate = LAMBDA1) + P2 * dexp(x_vals, rate = LAMBDA2)
lines(x_vals, densidad_teorica, col = "red", lwd = 2)

legend(
  "topright", 
  legend = c("Muestras Generadas", "Densidad Teórica"), 
  col = c("lightblue", "red"), 
  lty = c(NA, 1), 
  lwd = c(NA, 2), 
  pch = c(15, NA), 
  cex = 0.8
)

File: Unidad 3/Ejercicios/MetodoDeComposicion.py
import numpy as np
import matplotlib.pyplot as plt

def generar_composicion(n, p1, lambda1, lambda2):
   
    np.random.seed(42) 
    
    U = np.random.rand(n)
    
    X = np.empty(n)
    
    idx_comp1 = U < p1
    idx_comp2 = U >= p1
    
    n1 = np.sum(idx_comp1) 
    X[idx_comp1] = np.random.exponential(scale=1/lambda1, size=n1)
    
    n2 = np.sum(idx_comp2) 
    X[idx_comp2] = np.random.exponential(scale=1/lambda2, size=n2)
    
    return X

N = 10000          
P1 = 0.5          
LAMBDA1 = 0.5      
LAMBDA2 = 2        

muestras_comp = generar_composicion(N, P1, LAMBDA1, LAMBDA2)

print(f"Primeras 10 muestras generadas:\n{muestras_comp[:10]}\n")

plt.figure(figsize=(10, 6))

plt.hist(muestras_comp, bins=50, density=True, color='skyblue', edgecolor='black', alpha=0.7, label='Muestras Generadas')

x_vals = np.linspace(0, np.max(muestras_comp), 500)

densidad_teorica = P1 * (LAMBDA1 * np.exp(-LAMBDA1 * x_vals)) + \
                   (1 - P1) * (LAMBDA2 * np.exp(-LAMBDA2 * x_vals))

plt.plot(x_vals, densidad_teorica, color='red', linewidth=2, label='Densidad Teórica')

plt.title('Método de Composición: Mezcla de dos Exponenciales')
plt.xlabel('Valores de la Variable Aleatoria')
plt.ylabel('Densidad de Probabilidad')
plt.legend()
plt.grid(axis='y', alpha=0.5)
plt.show()